### PR TITLE
DR-1114: Add Test Runner Smoke Test to Int PR Github action

### PR DESCRIPTION
**Pull request by snf2ye**
* Add step to our integration test run on PR build to run Test Runner Smoke tests
* Add two env variables that allow the user to specify the service account used for delegating our test user's credentials (TEST_RUNNER_DELEGATOR_SA_FILE) and the service account used to run the rest of the Test Runner actions (TEST_RUNNER_SA_FILE). This is necessary because our credentials live in a slightly different place during the action run than during local test runs.

TODO: update github action version after merging [associated PR](https://github.com/broadinstitute/datarepo-actions/pull/26).

### CODE REVIEW

File: .github/workflows/gradle-build-pr.yml [hunk 1]
```diff
@@ -1,4 +1,8 @@
 name: Integration test and Connected tests
+env:
+  K8_CLUSTER: "integration-master"
+  GOOGLE_APPLICATION_CREDENTIALS: 'jade-dev-account.json'
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml:5]:
```suggestion
  # This must be defined for the bash redirection
  GOOGLE_APPLICATION_CREDENTIALS: 'jade-dev-account.json'
```

File: .github/workflows/gradle-build-pr.yml [hunk 2]
```diff
@@ -1,4 +1,8 @@
...
+  GOOGLE_SA_CERT: 'jade-dev-account.pem'
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml:7]:
```suggestion
  # This must be defined for the bash redirection
  GOOGLE_SA_CERT: 'jade-dev-account.pem'
```

File: .github/workflows/gradle-build-pr.yml [hunk 3]
```diff
@@ -27,7 +31,7 @@ jobs:
       - name: "Checkout code"
         uses: actions/checkout@v2
       - name: "Run Connected test via Gradle"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]: Change all these to what the new tag will be after the action is merged. 
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 4]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'gradlebuild'
       - name: "Deploy to cluster with Helm"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
         with:
           actions_subcommand: 'helmdeploy'
           helm_secret_chart_version: '0.0.4'
           helm_datarepo_api_chart_version: '0.0.17'
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]: adding chart up dates
```suggestion
          helm_datarepo_api_chart_version: '0.0.19'
```

File: .github/workflows/gradle-build-pr.yml [hunk 5]
```diff
@@ -57,43 +61,46 @@ jobs:
...
           helm_datarepo_ui_chart_version: '0.0.8'
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
          helm_datarepo_ui_chart_version: '0.0.9'
```

File: .github/workflows/gradle-build-pr.yml [hunk 6]
```diff
@@ -57,43 +61,46 @@ jobs:
       - name: "Checkout code"
         uses: actions/checkout@v2
       - name: "Whitelist Runner IP"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 7]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'gcp_whitelist'
           role_id: ${{ secrets.ROLE_ID }}
           secret_id: ${{ secrets.SECRET_ID }}
-          k8_cluster: 'integration-master'
       - name: "Check for an availble namespace to deploy API to and set state lock"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 8]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'k8_checknamespace'
-          k8_namespaces: 'integration-1,integration-2,integration-3'
+          k8_namespaces: 'integration-1,integration-2,integration-3, integration-6'
       - name: "Build docker container via Gradle"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 9]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'gradlebuild'
       - name: "Deploy to cluster with Helm"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 10]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'helmdeploy'
           helm_secret_chart_version: '0.0.4'
           helm_datarepo_api_chart_version: '0.0.17'
           helm_datarepo_ui_chart_version: '0.0.8'
           helm_gcloud_sqlproxy_chart_version: '0.19.7'
           helm_oidc_proxy_chart_version: '0.0.9'
+      - name: "Run Test Runner smoke tests via Gradle"
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 11]
```diff
@@ -57,43 +61,46 @@ jobs:
...
+        with:
+          actions_subcommand: 'gradletestrunnersmoketest'
       - name: "Run Integration test via Gradle"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 12]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'gradleinttest'
           pgport: ${{ job.services.postgres.ports[5432] }}
           test_to_run: 'testIntegration'
       - name: "Clean state lock from used Namespace on API deploy"
         if: always()
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 13]
```diff
@@ -57,43 +61,46 @@ jobs:
...
         with:
           actions_subcommand: 'k8_checknamespace_clean'
       - name: "Clean whitelisted Runner IP"
         if: always()
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
```
Comment (smark88) [inline .github/workflows/gradle-build-pr.yml]:
```suggestion
        uses: broadinstitute/datarepo-actions@0.20.0
```

File: .github/workflows/gradle-build-pr.yml [hunk 14]
```diff
@@ -57,43 +63,46 @@ jobs:
...
           role_id: ${{ secrets.ROLE_ID }}
           secret_id: ${{ secrets.SECRET_ID }}
-          k8_cluster: 'integration-master'
       - name: "Check for an availble namespace to deploy API to and set state lock"
-        uses: broadinstitute/datarepo-actions@0.14.0
+        uses: broadinstitute/datarepo-actions@sh-DR-1114-check-before-auth
         with:
           actions_subcommand: 'k8_checknamespace'
-          k8_namespaces: 'integration-1,integration-2,integration-3'
+          k8_namespaces: 'integration-1,integration-2,integration-3, integration-6'
```
Comment (marikomedlock) [inline .github/workflows/gradle-build-pr.yml:75]: Just curious, why did you add integration-6 here?

File: datarepo-clienttests/src/main/java/runner/config/TestUserSpecification.java [hunk 15]
```diff
@@ -31,12 +32,25 @@ public static TestUserSpecification fromJSONFile(String resourceFileName) throws
     TestUserSpecification testUser =
         objectMapper.readValue(inputStream, TestUserSpecification.class);
 
+    // read in the server file
+    String delegatorSaEnvVarOverride = readDelegatorSAFileEnvironmentVariable();
+    if (delegatorSaEnvVarOverride != null) {
+      testUser.delegatorServiceAccountFile = delegatorSaEnvVarOverride;
+    }
     testUser.delegatorServiceAccount =
         ServiceAccountSpecification.fromJSONFile(testUser.delegatorServiceAccountFile);
```
Comment (fboulnois) [inline datarepo-clienttests/src/main/java/runner/config/TestUserSpecification.java]:
```suggestion
    // override the service account if the environment variable is defined, otherwise use the test user one
    String delegatorSaEnvVarOverride = readDelegatorSAFileEnvironmentVariable();
    if (delegatorSaEnvVarOverride != null) {
      testUser.delegatorServiceAccountFile = delegatorSaEnvVarOverride;
      testUser.delegatorServiceAccount = ServiceAccountSpecification.fromJSONFile(delegatorSaEnvVarOverride);
    } else {
      testUser.delegatorServiceAccount = ServiceAccountSpecification.fromJSONFile(testUser.delegatorServiceAccountFile);
    }
```

Comment (fboulnois) [reply datarepo-clienttests/src/main/java/runner/config/TestUserSpecification.java]: This should be equivalent but I got confused between `testUser.delegatorServiceAccountFile` and `testUser.delegatorServiceAccount` since the variables look so similar. I also updated the comment.

**Review by fboulnois — APPROVED** [review]
one minor suggestion:

File: datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java [hunk 16]
```diff
@@ -90,6 +93,34 @@ public static GoogleCredentials getApplicationDefaultCredential() throws IOExcep
...
+      synchronized (lockTestRunnerSACredential) {
+        File jsonKey = testRunnerServiceAccount.jsonKeyFile;
+        testRunnerSACredential =
+            ServiceAccountCredentials.fromStream(new FileInputStream(jsonKey))
+                .createScoped(
+                    Collections.singletonList("https://www.googleapis.com/auth/cloud-platform"));
+      }
+      return testRunnerSACredential;
+    }
+    return getApplicationDefaultCredential();
```
Comment (fboulnois) [inline datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java:120]:
```suggestion
    // otherwise return the default `${GOOGLE_APPLICATION_CREDENTIALS}`
    return getApplicationDefaultCredential();
```

**Review by fboulnois — APPROVED** [review]
another small suggestion:

File: datarepo-clienttests/src/main/java/runner/config/TestUserSpecification.java (hunk 15 above)
Comment (snf2ye) [reply datarepo-clienttests/src/main/java/runner/config/TestUserSpecification.java]: Made the suggested change and added a more detailed comment - let me know if it makes sense.

File: datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java (hunk 16 above)
Comment (snf2ye) [reply datarepo-clienttests/src/main/java/common/utils/AuthenticationUtils.java:120]: Added a detailed comment - let me know if it makes sense.

**Comment by fboulnois** [comment]
Nice I like the comments!

**Review by fboulnois — APPROVED** [review]
nice I like the comments, lgtm

File: datarepo-clienttests/src/main/resources/suites/PRSmokeTests.json [hunk 17]
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "PRSmokeTests",
+  "description": "Two smoke tests to run on PR build. One requires authentication and one does not. Only queries, no data creation.",
```
Comment (marikomedlock) [inline datarepo-clienttests/src/main/resources/suites/PRSmokeTests.json:3]: Perhaps update description to include the LookupDataset config?

**Review by marikomedlock — APPROVED** [review]
LGTM. I think we need the follow-on changes to the authentication stuff, so I'll put that in my pending PR.

**Comment by snf2ye** [comment]
int test failure: https://broadworkbench.atlassian.net/browse/DR-1043
